src/Language/PureScript/Make.hs

```diff
@@ -261,6 +263,46 @@ make ma@MakeActions{..} ms = do
   onExceptionLifted :: m a -> m b -> m a
   onExceptionLifted l r = control $ \runInIO -> runInIO l `onException` runInIO r
 
+  -- Remove missing and duplicate files from the cache.
+  -- This action removes modules if their source does not exist or if another
+  -- module with the same source is being compiled.
+  pruneCache :: S.Set ModuleName -> Cache.CacheDb -> m Cache.CacheDb
+  pruneCache modules cache =
+    let
+      inverseCache :: M.Map FilePath [ModuleName]
+      inverseCache =
+        M.foldlWithKey
+          (\acc1 key e -> M.unionWith (++) acc1 $ foldl (\acc2 file -> M.insert file [key] acc2) M.empty (M.keys (Cache.unCacheInfo e)))
+          M.empty
+          cache
+
+      -- Set of modules that should be removed due to there being two or more modules with the same source
+      toRemove :: S.Set ModuleName
+      toRemove =
+        M.foldl
+          (\acc1 e -> case e of
+            -- This pattern will clean up duplicates over time.
+            -- It will keep the cache in a good state if it was in a good state before.
+            m1 : m2 : _ ->
```

Comment:
isn't this just when `Set.size e > 1` then add `e \\ modules` to the accumulator? Then there's no need to wait until multiple compilations reach a fixed point (assuming I understand the comment)

Reply:
I don't think so? What this means is if there are two or more modules with the same name in the cache, at most one of them is being compiled right now (i.e. it is in the modules set). `e \\ modules` (assuming \\\\ is set difference) will be all duplicate modules, but we want to ignore those, not find them. Perhaps `e \\ (e \\ modules)` is easier to understand?

Anyway, in practice it is possible there are multiple duplicates if you rename and compile several times, but after this change it is only possible to have a single duplicate, so this code is enough. If you think it is better to refactor it I will do that

Reply:
Sorry, not sure if I was unclear or I don't understand - I was suggesting that `e \\ modules` is added to the accumulator that goes into this fold to construct `toRemove`, as `acc2 = S.union (Set.singleton m1 \\ modules) acc1` etc

I don't feel strongly about the code, just trying to follow through and understand "clean up duplicates over time"

Reply:
I think I understand what you mean now. I must have been tired when I read your comment initially :) Yes, just doing the set difference is probably the way to go. I will refactor it

Reply:
While you're in there, I think this might be a bit clearer if you separate out the duplicate pruning and the missing file pruning into separate functions.

Also, general tip: you're doing a lot of bespoke folds here to accumulate and aggregate stuff, and I think you'd be better off using simpler primitives like `concat` and `filter` and `M.fromListWith`.